src/view3d.cpp

```diff
@@ -185,61 +209,156 @@ void View3D::centerOnGeometry(const QModelIndex& index)
                   (const char*)glGetString(GL_RENDERER),
-                  (const char*)glGetString(GL_VERSION));
-    m_shaderProgram->setContext(context());
-    m_meshFaceShader.reset(new ShaderProgram(context()));
+                  (const char*)glGetString(GL_VERSION),
+                  (const char*)glGetString(GL_SHADING_LANGUAGE_VERSION),
+                  (const char*)glewGetString(GLEW_VERSION));
+
+    // GL_CHECK has to be defined for this to actually do something
+    glCheckError();
+
+    initCursor(10, 1);
+    initAxes();
+    initGrid(2.0f);
+
+    glCheckError();
+
+    m_boundingBoxShader.reset(new ShaderProgram());
+    m_boundingBoxShader->setShaderFromSourceFile("shaders:bounding_box.glsl");
+
+    m_meshFaceShader.reset(new ShaderProgram());
     m_meshFaceShader->setShaderFromSourceFile("shaders:meshface.glsl");
-    m_meshEdgeShader.reset(new ShaderProgram(context()));
+
+    m_meshEdgeShader.reset(new ShaderProgram());
     m_meshEdgeShader->setShaderFromSourceFile("shaders:meshedge.glsl");
-    m_incrementalFramebuffer = allocIncrementalFramebuffer(width(), height());
+
+    double dPR = getDevicePixelRatio();
+    int w = width() * dPR;
+    int h = height() * dPR;
+
+    m_incrementalFramebuffer = allocIncrementalFramebuffer(w, h);
+
+    glFrameBufferStatus(m_incrementalFramebuffer);
+    glCheckError();
+
     const GeometryCollection::GeometryVec& geoms = m_geometries->get();
     for (size_t i = 0; i < geoms.size(); ++i)
-        geoms[i]->initializeGL();
+    {
+        if (m_boundingBoxShader->isValid())
+        {
+            // TODO: build a shader manager for this
+            geoms[i]->setShaderId("boundingbox", m_boundingBoxShader->shaderProgram().programId());
+            geoms[i]->setShaderId("meshface", m_meshFaceShader->shaderProgram().programId());
+            geoms[i]->setShaderId("meshedge", m_meshEdgeShader->shaderProgram().programId());
+            geoms[i]->initializeGL();
+        }
+    }
+
+    PointViewerMainWindow * pv_parent = dynamic_cast<PointViewerMainWindow *>(parentWidget());
+    pv_parent->openShaderFile("shaders:las_points.glsl");
+
+    setFocus();
+
+    // resizeGL(w,h);
+    // updateGL();
 }
 
 
 void View3D::resizeGL(int w, int h)
 {
+    //Note: This function receives "device pixel sizes" for correct OpenGL viewport setup
+    //      Under OS X with retina display, this becomes important as it is 2x the width() or height()
+    //      of the normal window (there is a devicePixelRatio() function to deal with this).
+
     if (m_badOpenGL)
         return;
+
     // Draw on full window
     glViewport(0, 0, w, h);
-    m_camera.setViewport(QRect(0,0,w,h));
+
+    double dPR = getDevicePixelRatio();
+
+    m_camera.setViewport(QRect(0,0,double(w)/dPR,double(h)/dPR));
+
     m_incrementalFramebuffer = allocIncrementalFramebuffer(w,h);
+
+    glCheckError();
+
+    // without this explicit call, we get very strange artifacts during resizing
+    //updateGL();
```

Comment:
This was true, while using the Qt Framebuffer and I didn't see this problem anymore after switching to native OpenGL calls.